sdk/python/kfp/components/structures/components.json_schema.outline.yaml

```diff
@@ -0,0 +1,77 @@
+<ComponentSpec>:
+    name: string
+    description: string
+    source: #SourceSpec
+        url: string
+    schemaVersion: string
+    inputs: #InputSpec[]
+        <InputSpec[]>:
+            name: string
+            type: TypeType
+            description: string
+            default: PrimitiveTypes
+            optional: boolean
+    outputs: #OutputSpec[]
+        <OutputSpec[]>:
+            name: string
+            type: TypeType
+            description: string
+    implementation: #ImplementationType
+        <oneOf>:
+            <ContainerImplementation>:
+                container: #ContainerSpec
+                    image: string
+                    command: #CommandlineArgumentType
+                        <oneOf>:
+                            <PrimitiveTypes>:
+                                <oneOf>: string, integer, number, boolean
+                            <InputValuePlaceholder>:
+                                inputValue: string
+                            <InputPathPlaceholder>:
```

Comment:
Can you give more context on InputPathPlaceholder and OutputPathPlaceholder?

Reply:
+1

Reply:
Quoting from the full schema:

`InputValuePlaceholder`:  Represents the command-line argument placeholder that will be replaced at run-time by the input argument value.

Argo:

    command: [program.py, "{{inputs.parameters.input_name}}"]

Component YAML:

    command: [program.py, {inputValue: Input name}]

In Argo, placeholders are not part of the schema. Argo is using ad-hoc string replacement which in some cases leaves the placeholder not replaced.

Unlike in Argo, in component.yaml the placeholders are part of the schema: they're strongly typed and are validated.

`InputPathPlaceholder`: Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file containing the input argument value.

`OutputPathPlaceholder`: Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file where the program should write its output data.

Input or output path placeholders are replaced by a system-generated local file path by the pipeline system during the compilation. System control over input/output paths is required for advanced storage scenarios like fast volume-based artifact passing, intermediate data caching, etc.

For a typical program the command-line will be resolved the following way:
In component.yaml:

    command: [train.py, --train-input, {inputPath: Train set}, --output-model, {outputPath: Model}, --rounds, {inputValue: Rounds}]

Resolved:

    command: [train.py, --train-input, /outputs/train_set/data, --output-model, /outputs/model/data, --rounds, 100]

I can add some more examples if needed.